### DISCUSSION TITLE
BOLD system made their own package. Is this one still needed?

### DISCUSSION THREAD

**Discussion by salix-d** [ropensci/bold#112] [General]
Hello everyone,

I just learned that [BOLD system](https://boldsystems.org) made their own [r package](https://github.com/boldsystems-central/BOLDconnectR/).

It seems to have more a lot more functionalities, including the ability to fetch of private records and analyse the data.
From my quick look on it, the way to fetch the data is different from this package as it reflects the structure of their new API (v5).
I haven't try getting taxonomy data with their package yet, and the website indicates it's still using the v4 API though, which this package uses.

So, if you could try out their package and give your feedback from it (comparing it with this one), I would greatly appreciate it.
I would mainly like to know if their package covers all your needs or if there are still functionalities this package is better suited for.
With that feedback, I'll be able to plan what to do next.

Thank you!

**Comment by mpadge** [comment]
@salix-d This is not an answer to your question, but might nevertheless give you some useful insight. The following graphs come from https://github.com/ropensci-review-tools/repometrics, and running:
``` r
data <- repometrics_data("/<local>/<path>/<to>/bold")
repometrics_dashboard(data)
```

### Number of new issues opened

![image](https://github.com/user-attachments/assets/4945672a-0ed5-4541-b72e-3037733b8808)

### Number of issue comments

![image](https://github.com/user-attachments/assets/92238aa8-7f34-4b4f-80bc-f35db08f7cf6)

They both indicate a gradual decline, and so might plausibly be interpreted to support a decision to archive this pkg?

---

Footnote: That `repometrics` pkg is a very rapidly changing WIP, and currently just demonstrates the kinds of things possible. No guarantee of ability to reproduce those results in the future ...)

**Reply by salix-d** [reply]
I thought it meant the package was doing well without issues 😅

Thanks for the tool!

**Comment by noamross** [comment]
This kind of thing is always challenging.  Despite what I think in an increase focus in OSS and R on getting people to look at prior art, it's very common for competing packages to spring up and very rare for projects to coordinate/merge.  And that's not neccessarily bad! Sometimes there are different needs and both should be maintained and/or it makes sense for there to be succession.  But I'd encourage you (if you haven't already), to reach out the BOLDconnectR team to discuss possible collaboration. As {bold} has been around 14 years and has a number of reverse dependencies in the taxonomic ecosystem, there are probably lessons to be learned. Maybe it makes sense to integrate some components, maybe they can give insight as to expected API migration, or maybe they just want to do their own thing and that's fine.  Given the timing, it's also possible that the package coming off CRAN was part of their motivation, and it would be good to discuss, as well.

**Reply by salix-d** [reply]
I think the reason for it stems more from their new API; the structure is pretty different.
They were aware of this package and that I was its maintainer; will still try to discuss.

Re: different needs:
That's exacly why I've opened this discussion, to see if that was the case.

**Comment by RobLBaker** [comment]
For packages that do have dependencies on `bold` - or depend on packages that themselves depend on `bold` - what are your suggestions? Is there a timeline for either getting `bold` back on CRAN or a different approach? If you'd like help, I'm happy to lend a hand towards whatever the most workable solution ends up being.

**Reply by salix-d** [reply]
Hi! 

I know some packages that depends on github only packages and they simply add the instructions on how to install from github (from what I've seen, there might be more under the hood I haven't noticed though). 

But I'm trying to get it back on CRAN, I'm not exactly sure what went wrong with my last submission. I have another one almost ready. If that one doesn't pass, I would appreciate some help with that.

Although, even if it passes, the problem is that it will eventually fail as the version of the API used will not be maintained anymore.

Hence, this conversation to see if the new package made by BOLD system (BOLDconnectR) is a viable alternative now.

**Reply by salix-d** [reply]
If it's not a viable option, then I could use some help figuring out the API version switch. It seem like it will impact how I fetch the data a lot.

**Reply by RobLBaker** [reply]
Happy to help out! I have to admit I have very little experience with getting anything on CRAN, but I've also always wanted an excuse to learn more about it.

**Comment by salix-d** [comment]
re https://github.com/ropensci/bold/issues/110#issuecomment-2545445269

> My conclusion is that, at the moment, it's practically impossible to get hold of data at scale from BOLD

Well, that's a bummer, I though you only needed the API key for private data, I was able to get some data without it.

Did you try replying with "to use your r package" to their "What do you want to use the API key for?" question? (I would assume so, but just wanna make sure I  have all the info)

> However... I'm continuing to get a wide range of failures from using this library too.

Oh that's not good :(
I'll try your script to see where those failure are.

**Comment by boopsboops** [comment]
Hi Salix,

I realise I'm a bit late to the discussion with this ... but for what it's worth I have a pipeline which uses `ropensci/bold` heavily.  It recently broke, I assume because they have now taken v4 API offline?

I've just had a chance to test out `BOLDconnectR`. Immediately I ran into issues as it had 80 dependencies, many of which are geospatial packages which can be a pain to install if certain libraries are not up to date. 

After that was sorted, I passed `BOLDconnectR::bold.public.search()`  a list of 30 species to test the search on. The search failed in entirety because some of these species did not have any public data available on BOLD, which is stated in the function help: "Misspelled queries or those for which no public data exists on BOLD at the time the function is executed will result in an error."

You obviously can't know in advance which species have data and which don't, so it looks like the use case of searching on a species list is no longer possible. Seems like quite basic functionality. 

The solution is to execute an independent search for each species? I can't imagine this working as my IP address was blocked by BOLD after about 30 mins of me messing around testing on a handful of very small searches. 

So, if there's any possibility of resurrecting `ropensci/bold` for v5 API, I'd really appreciate that! Happy to help where I can.

Rupert

**Comment by salix-d** [comment]
Thanks! that's what I wanted to know.

> Immediately I ran into issues as it had 80 dependencies

Yeah, I saw that and was shocked. A lot are needed only when doing specific analysis, they should've went into 'recommend' with a warning from functions using it.

> "Misspelled queries or those for which no public data exists on BOLD at the time the function is executed will result in an error."

Ooof... such bad error handling...

> my IP address was blocked

Permenently or temporarily?

> So, if there's any possibility of resurrecting ropensci/bold for v5 API, I'd really appreciate that! Happy to help where I can.

I'll invetigate this. They don't seem to have specific API documentation like they use to since they probably expect users to use their website or package directly. Might be able to infer it from their source code though 🤔 so I'll have a look at that.

**Comment by salix-d** [comment]
@boopsboops Could I have your list of species to test with as well?

Also, rebuilding this is gonna take me a while as I'm not as available as I was previously. 

You might want to try their API directly. It's actually well documented -- I was looking for it in the wrong place when I last commented (it's under 'Data' and not 'Ressource Hub').

[APIs page](https://boldsystems.org/data/api/): careful API button at top and bottom of page don't link to  the same API. First is data, second is portal. I think for species/records information you want [portal](https://portal.boldsystems.org/api/docs)

**Comment by salix-d** [comment]
@boopsboops I also think it would be worth rising those issues on their [github issue's page](https://github.com/boldsystems-central/BOLDconnectR/issues).
